include/oneapi/dpl/experimental/kt/internal/esimd_radix_sort_kernels.h

```diff
@@ -0,0 +1,706 @@
+#include "../../../pstl/utils.h"
+
+#include "esimd_defs.h"
+#include "esimd_radix_sort_utils.h"
+
+namespace oneapi::dpl::experimental::kt::esimd::__impl
+{
+
+template <bool __is_ascending, ::std::uint8_t __radix_bits, ::std::uint16_t __data_per_work_item,
+          ::std::uint16_t __work_group_size, typename _KeyT, typename _RngPack>
+_ONEDPL_ESIMD_INLINE void
+__one_wg_kernel(sycl::nd_item<1> __idx, ::std::uint32_t __n, _RngPack&& __rng_pack)
+{
+    using _BinT = ::std::uint16_t;
+    using _HistT = ::std::uint16_t;
+    using _DeviceAddrT = ::std::uint32_t;
+
+    constexpr ::std::uint32_t __bin_count = 1 << __radix_bits;
+    constexpr ::std::uint32_t __bit_count = sizeof(_KeyT) * 8;
+    constexpr ::std::uint32_t __stage_count = oneapi::dpl::__internal::__dpl_ceiling_div(__bit_count, __radix_bits);
+    constexpr _BinT __mask = __bin_count - 1;
+    constexpr ::std::uint32_t __hist_stride = sizeof(_HistT) * __bin_count;
+
+    constexpr ::std::uint32_t __reorder_slm_size = __data_per_work_item * sizeof(_KeyT) * __work_group_size;
+    constexpr ::std::uint32_t __bin_hist_slm_size = __hist_stride * __work_group_size;
+    constexpr ::std::uint32_t __incoming_offset_slm_size = (__bin_count + 1) * sizeof(_HistT);
+
+    // max SLM is 256 * 4 * 64 + 256 * 2 * 64 + 257*2, 97KB, when  __data_per_work_item = 256, __bin_count = 256
+    // to support 512 processing size, we can use all SLM as reorder buffer with cost of more barrier
+    __dpl_esimd::__ns::slm_init<::std::max(__reorder_slm_size, __bin_hist_slm_size + __incoming_offset_slm_size)>();
+
+    const ::std::uint32_t __local_tid = __idx.get_local_linear_id();
+
+    const ::std::uint32_t __slm_reorder_this_thread = __local_tid * __data_per_work_item * sizeof(_KeyT);
+    const ::std::uint32_t __slm_bin_hist_this_thread = __local_tid * __hist_stride;
+
+    __dpl_esimd::__ns::simd<_HistT, __bin_count> __bin_offset;
+    __dpl_esimd::__ns::simd<_DeviceAddrT, __data_per_work_item> __write_addr;
+    __dpl_esimd::__ns::simd<_KeyT, __data_per_work_item> __keys;
+    __dpl_esimd::__ns::simd<_BinT, __data_per_work_item> __bins;
+    __dpl_esimd::__ns::simd<_DeviceAddrT, __data_per_step> __lane_id(0, 1);
+
+    const _DeviceAddrT __io_offset = __data_per_work_item * __local_tid;
+
+    static_assert(__data_per_work_item % __data_per_step == 0);
+    static_assert(__bin_count % 128 == 0);
+    static_assert(__bin_count % 32 == 0);
+    _ONEDPL_PRAGMA_UNROLL
+    for (::std::uint32_t __s = 0; __s < __data_per_work_item; __s += __data_per_step)
+    {
+        __dpl_esimd::__ns::simd_mask<__data_per_step> __m = (__io_offset + __lane_id + __s) < __n;
+        __keys.template select<__data_per_step, 1>(__s) =
+            __dpl_esimd::__ns::merge(__dpl_esimd::__gather<_KeyT, __data_per_step>(__rng_data(__rng_pack.__keys_rng()), __lane_id, __io_offset + __s, __m),
+                  __dpl_esimd::__ns::simd<_KeyT, __data_per_step>(__sort_identity<_KeyT, __is_ascending>()), __m);
```

Comment:
Is `__gather` required here to on all input because incoming data user may not be aligned to our needs?
If so, aren't we paying a huge price to align it by gathering across all input data?  

I think we could probably get away with a single gather.

We are already padding at the end for when `__n` and a multiple of `__data_per_work_item * __work_group_size` do not align.  Can we also pad at the start with `__sort_identity<_KeyT, **!__is_ascending**>` to keep those elements at the beginning of the sorted list? 

I think this would allow us to special case this first load to get to the start of the aligned input data, and then have we can skip the gather for the rest of the copy-in.  

The same idea would be applied on the way out for the scatter.  Special case the first step, then deal in aligned data the rest of the way.

Reply:
> Is __gather required here to on all input because incoming data user may not be aligned to our needs?

Yes, but this is not entirely correct. User-provided will always be element-aligned (otherwise, the user does not follow an inherent C++ requirement on the alignment). `gather` only requires the passed offset to be element-aligned. However, there are other, potentially more efficient, instructions with no overalignment requirements.

E.g. [block_load](https://intel.github.io/llvm-docs/doxygen/group__sycl__esimd__memory__block.html#ga49ae87e4075911a7e92549fb1e25fdb8), which may be used to load non-resudial elements. The API reference is not very clear so far, but my understanding is that an offset passed to `block_load` must be at least element-aligned if you provide the alignment property. 
 
> I think we could probably get away with a single gather (or maybe none?).

We can  reuse `__load_simd` wrapper over `gather` from onesweep kernel. It also applies the mask only to the residual elements, which is potentially more performant.

I think, we still need to load data in a loop portion by portion if we use `gather`. Single use with large width may increase register pressure for keeping a long mask vector.

Reply:
> We are already padding at the end for when __n and a multiple of __data_per_work_item * __work_group_size do not align. 

This is not for the sake of alignment. The "sort identity" elements are added to avoid extra branching. They will always be at the end of the sequence after reordering. So, you just need to avoid scattering the last elements; no need to add extra `if` conditions in the code.

Reply:
This is noted in #1312. Thanks for the great comment!